lib/jobs/ServiceInstanceUpdateJob.js

```diff
@@ -0,0 +1,265 @@
+      throw new errors.Forbidden(`Automatic update not possible. Detected changes in forbidden section(s) '${forbiddenSections.join(',')}'`);
+    }
+    return false;
+  }
+
+  static rescheduleUpdateJob(instanceDetails, trackAttempts) {
+    return Promise.try(() => {
+      const jobData = _.cloneDeep(instanceDetails);
+      jobData.attempt = jobData.attempt + 1;
+      if (trackAttempts) {
+        if (jobData.attempt > config.scheduler.jobs.service_instance_update.max_attempts) {
+          logger.error(`Auto udpate for instance ${instanceDetails.instance_id}  has exceeded max configured attempts : ${config.scheduler.jobs.service_instance_update.max_attempts}}`);
+          return;
+        }
+      }
+      logger.info(`Schedulding InstanceUpdate Job for ${instanceDetails.instance_name}:${instanceDetails.instance_id} @ ${config.scheduler.jobs.reschedule_delay} - Track : ${trackAttempts} - Attempt - ${instanceDetails.attempt}`);
+      return ScheduleManager
+        .runAt(instanceDetails.instance_id,
+          CONST.JOB.SERVICE_INSTANCE_UPDATE,
+          config.scheduler.jobs.reschedule_delay,
+          jobData,
+          CONST.SYSTEM_USER
+        );
+    });
+  }
+
+  static getRandomRepeatInterval() {
+    const afterHr = _.get(config, 'scheduler.jobs.service_instance_update.should_start_after_hr', 0);
+    const beforeHr = _.get(config, 'scheduler.jobs.service_instance_update.should_start_before_hr', 23);
+    const afterMin = _.get(config, 'scheduler.jobs.service_instance_update.should_start_after_min', 0);
+    const beforeMin = _.get(config, 'scheduler.jobs.service_instance_update.should_start_before_min', 59);
+    const runOnceEvery = _.get(config, 'scheduler.jobs.service_instance_update.run_every_xdays', 15);
+    const opts = {
+      start_after_hr: afterHr,
+      start_before_hr: beforeHr,
+      start_after_min: afterMin,
+      start_before_min: beforeMin
+    };
+    return utils.getRandomCronForOnceEveryXDays(runOnceEvery, opts);
+  }
+
+  static getLastRunStatus(name) {
+    const sortOn = ['createdAt', 'asc'];
+    const criteria = {
+      sortBy: [sortOn],
+      searchBy: {
+        name: name,
+        type: CONST.JOB.SERVICE_INSTANCE_UPDATE
+      }
+    };
+    return Repository
+      .search(CONST.DB_MODEL.JOB_RUN_DETAIL,
+        criteria, {
+          records: config.scheduler.jobs.service_instance_update.max_attempts,
+          offset: 0
+        })
+      .then(lastRunDetails => {
+        logger.info('LastRun details retrieved - ', lastRunDetails);
+        if (_.get(lastRunDetails, 'totalRecordCount', 0) === 0) {
+          return null;
+        }
+        const lastRunList = lastRunDetails.list;
+        const count = lastRunList.length;
+        let initialJobRunStatus, idx;
+        for (idx = count - 1; idx >= 0; idx--) {
+          if (lastRunList[idx].data.attempt === 1) {
+            initialJobRunStatus = lastRunList[idx];
```

Comment:
since the sort option is asc above initialjobrunstatus will be at index 0 , correct ?

Reply:
changed sort order per above comment. Also we are fetching last 3 run history records we cant always assume idx = 0 (asc) or idx = 2 (desc) to be initial run status, because many a times the update might complete with 1 attempt, in which case at idx=1, initialJobStatus will be found.